chips/earlgrey/src/pinmux.rs

```diff
@@ -0,0 +1,208 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2022.
+
+//! Mux'ing between physical pads and GPIO or other peripherals.
+
+use kernel::hil::gpio;
+use kernel::utilities::registers::interfaces::{Readable, Writeable};
+use kernel::utilities::registers::{register_bitfields, FieldValue, LocalRegisterCopy};
+use kernel::utilities::StaticRef;
+
+use crate::registers::pinmux_regs::{
+    PinmuxRegisters, DIO_PAD_ATTR_REGWEN, MIO_OUTSEL_REGWEN, MIO_PAD_ATTR_REGWEN,
+    MIO_PERIPH_INSEL_REGWEN,
+};
+use crate::registers::top_earlgrey::{
+    DirectPads, MuxedPads, PinmuxInsel, PinmuxOutsel, PinmuxPeripheralIn, PINMUX_AON_BASE_ADDR,
+};
+
+pub const PINMUX_BASE: StaticRef<PinmuxRegisters> =
+    unsafe { StaticRef::new(PINMUX_AON_BASE_ADDR as *const PinmuxRegisters) };
+
+// To avoid code duplication for MIO/DIO we introduce
+// one register layout for both types of IO. In the future this code
+// should be replaced by official improved auto generated definitions.
+// OpenTitan documentation reference:
+// https://opentitan.org/book/hw/ip/pinmux/doc/registers.html#fields-6
+// https://opentitan.org/book/hw/ip/pinmux/doc/registers.html#fields-8
+register_bitfields![u32,
+    pub(crate) PAD_ATTR [
+        INVERT OFFSET(0) NUMBITS(1) [],
+        VIRTUAL_OPEN_DRAIN_EN OFFSET(1) NUMBITS(1) [],
+        PULL_EN OFFSET(2) NUMBITS(1) [],
+        PULL OFFSET(3) NUMBITS(1) [
+            DOWN = 0,
+            UP = 1,
+        ],
+        KEEPER_EN OFFSET(4) NUMBITS(1) [],
+        SCHMITT_EN OFFSET(5) NUMBITS(1) [],
+        OPEN_DRAIN_EN OFFSET(6) NUMBITS(1) [],
+        SLEW_RATE OFFSET(16) NUMBITS(2) [],
+        DRIVE_STRENGTH OFFSET(20) NUMBITS(4) [],
+    ],
+];
+
+type PadAttribute = LocalRegisterCopy<u32, PAD_ATTR::Register>;
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum Pad {
+    Mio(MuxedPads),
+    Dio(DirectPads),
+}
+
+impl Pad {
```

Comment:
Can we  use the `gpio::Configure` HIL here instead?

Reply:
You need information from mutex selection for making a working GPIO pin.
I think the best pace to aggregate all configurations is GPIO struct.

Reply:
Do you mean the current GPIO implementation will call these?

Reply:
Yes

Reply:
Ah ok. I agree that can be done separately. That is a useful bit of information to include in the commit message next time as it sets the context for the change

Reply:
Ah, I do see it under TODO in the description

Generally TODO applies to things that need to be done in this PR before it is ready to be merged